### DISCUSSION TITLE
Is this change for doesntExistOr(...), to whenExists(...), a worthwhile change?

### DISCUSSION THREAD

**Discussion by Barry20** [laravel/framework#46736] [General]
I was about to create an update to have new methods on the Query\Builder class, whenExists($closure) and whenDoesntExist($closure), the premise is simple, run a closure when the exists test is true. The main reason I want to do this is code readability. e.g.

```
MyModel::where('column', 'value')
    ->whenExists(function () {
        do_stuff();
    });
```

At this point I noticed existsOr($closure) and doesntExistOr($closure), which is effectively the same, except it has a mixed return value. I feel like the naming of whenExists() is more readable, it's closer to natural language and it's better targeted at the focal point, the closure, the existing name reads like it's inverse, i.e. to run the closure when records exists use doesntExistOr method.

So, my questions to the community are these, is it worth a pull request to rename this method and change return value to void? Or, if the existing context is worthwhile keeping, is it worth creating totally new methods specific to this purpose with the different naming and void return?

Ps. clean and succinct code is the goal, so removing the need for - if (...->exists()) {...} and helping readability.

**Comment by Angel5a** [comment]
Both existsOr($closure) and doesntExistOr($closure) are useful for their ability to return value from the callback, so you can put extra check inside callback:
```
$canProceed = MyModel::where('column', 'value')
    ->existsOr(fn () => $this->tryToPutNewRecord());
```
Yes, we can get the same result with a:
```
$canProceed = MyModel::where('column', 'value')->exists() || $this->tryToPutNewRecord();
```
But what about that "readability"?

Why void? Wouldn't it be better to return some CheckExistenceResult object, to chain actions:
```
MyModel::where('column', 'value')
    ->whenExists(function () {
        do_stuff();
    })
    ->whenDoesntExist(function () {
        do_other_stuff();
    });
```
So removing the need for - if (...->exists()) {...} else {...}?

I can't say, what would be better: to bloat api or to make BC. How often people use existsOr?

**Reply by Barry20** [reply]
Hmm, yes the chaining is a nice way to open it up and make it more flexible. I like that you could do something like:

```
MyModel::where('column', 'value')
    ->whenDoesntExist(function () {
        // create the model
        // do some other stuff in relation to this
    })
    ->first(); // now we have the MyModel with the other stuff performed
```

After my initial post, I did think of some use cases to warrant keeping the existsOr(), which I assume is the original intent or adding secondary checks, e.g. "Account User `existsOr` Account has a manager User".

I guess then yes, the question becomes simply, is it bloat? I don't think so but am open to thoughts.

**Reply by Angel5a** [reply]
Exists request doesn't provide any records. Databases have same optimizations to make this query runs faster. So making possible to call `first()` is not so good idea - if you need a model, you can simple call to `firstOr()`.

Two extra methods - not a big deal. We're already have a bunch of them.
